### ISSUE TITLE
Document performance surprises in Leo's API

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#464]
The huge performance bugs in #456 were the indirect result of several problems in Leo's API.

**Problem 1**: `p.b = x` does a lot more than advertised. This is confusing and has serious performance implications. Naively, one would expect `p.b = x` simply to call `p.v.setBodyString()`.  But no, the `p.b` setter property is equivalent to `c.setBodyString(p)`.

**`c.setBodyString is expensive!`**  Depending on circumstances, it can call `p.setDirty()`, `c.recolor()` and `c.redraw_after_icons_changed()`.

cff in leoPy.leo shows there are 55 matches for `.b = x` and 38 matches for `setBodyString`. There is rampant confusion in the code. 

**Problem 2**: `p.setDirty()` is much more expensive than `p.v.setDirty()`.  `p.setDirty()` calls `p.setAllAncestorAtFileNodesDirty()`.

**Problem 3**: `p.setHeadString()` calls `p.setDirty()`, which is another performance surprise.

The general rule, broken in all these cases, is that commander methods are "high level", meaning that they can redraw the screen and do other expensive things.  Furthermore, we want position properties and vnode properties to be uniformly fast. The solution to this issue is to honor that rule as follows:

**Change 1**:  Deprecate the `p.b` setter property.  Replace all `p.b = x` instances by calls to `c.setBody(p, x)`, where `c.setBody(p, x)` is a synonym for `c.setBodyString(p, x)`. This must be done throughout Leo, including all official plugins. Replace `c.setBodyString` by the preferred `c.setBody`.

**Change 2**: Deprecate `p.setDirty()`. Replace all calls by calls to a new method, `c.setDirty(p)`.

**Change 3**: Deprecate `p.setHeadString(x)`. Replace all calls by calls to a new method, `c.setHead(p, x)`.

**Change 4**: ~~All newly-deprecated properties and methods should issue a deprecation warning, created by the new function, `g.deprecated(name)`.~~

On second thought, the `p.b` setter and the `p.setDirty()` method should simply do the obvious things, namely call the corresponding VNode method. Still dithering about p.setHeadString().  Just deleting it may be good enough.

If `g.deprecated` *is* needed, it should maintain a dictionary so that deprecation warnings are given only once.

**Comment by tbnorth** [comment]
Hmm, tricky stuff.  Not super keen on c.setHead(p, ) replacing p.h, that latter's so much neater shorthand.

Accepting that the current condition is confusion, seems there's a couple of directions we could go.  It's always been and I think will continue to be the case that you should call c.redraw() at the end of script.  Part of the current confused state is that for some p.x() operations you don't need to. So on the one hand we could just make p.x() performant, i.e. vnode proxies, and everything should work ok as long as people call c.redraw(), and maybe c.setDirty() or something.

OTOH p.v.x() are already the performant versions, we could just document that fact and use those.

This is assuming (3) suggests removing the p.h setter, I guess you don't say that explicitly.

Cheers -Terry

**Comment by tbnorth** [comment]
See #465 for tangentially related performance discussion.

**Comment by edreamleo** [comment]
​​
On Sun, Apr 2, 2017 at 2:58 PM, Terry Brown <notifications@github.com>
wrote:

> Hmm, tricky stuff. Not super keen on c.setHead(p, ) replacing p.h, that
> latter's so much neater shorthand.
>
​After messing around in the api branch for several hours, I am tending to
agree with you.​

Besides being a super abbreviation visually, the p.b setter automatically
gets c from p.context.

So I am inclined to leave p.b as it is.  In fact, I'm inclined to leave
*everything* as it is.

> Accepting that the current condition is confusion, seems there's a couple
> of directions we could go. It's always been and I think will continue to be
> the case that you should call c.redraw() at the end of script. Part of
> ​​
> the current confused state is that for some p.x() operations you don't
> need to.
>
> ​The p.b setter calls c.setBodyString(s), which might call c.recolor(),
but never calls c.redraw(). We only need to call c.redraw() if nodes are
inserted or deleted.

> So on the one hand we could just make p.x() performant, i.e. vnode
> proxies, and everything should work ok as long as people call c.redraw(),
> and maybe c.setDirty() or something.
>
​Alas not.  c.setBodyString(s) does several tricky things that c.redraw()
doesn't do.

> OTOH p.v.x() are already the performant versions, we could just document
> that fact and use those.
>
> This is assuming (3) suggests removing the p.h setter, I guess you don't
> say that explicitly.
>
Meh, I just realized that p.h calls p.setDirty(), so it too has the
potential to ruin performance.

Happily, serious performance problems can only occur in highly repetitive
operations that *change* headline or body text:

- cff has always been very fast.

- replace-all is now 100x faster than before. As the result of this
discussion I'll check p.h setters in that code.

- The RecursiveImportController class might be greatly accelerated by
replacing p.b = x by p.b.v = x. We need p.b = x only if the body text
is *immediately
*going to become visible. c.p.b might have to be "refreshed" somehow at the
end of the import. We'll see...

To summarize:

- Big performance bugs can only arise in highly repetitive operations that
*change* or *create* text or nodes.  replace-all and cff are (now) very
fast.  recursive import can probably be greatly accelerated.

- The p.b and p.h setters are *much* slower than the p.v.b and p.v.h
setters.  However, the p setters are generally necessary, because they
update the UI in complex ways.

- Similarly, p.setDirty() is *much* slower than p.v.setDirty(), but is
often necessary.

- c.redraw() is not a substitute for the UI operations done by the p.b and
p.b setters and by p.setDirty().

- At present, I am inclined to keep everything as it is, except for any
*local* changes that a thorough code review may suggest.

Edward

**Comment by xgid** [comment]
I have just started to look at the API and so far I've made only a small script, so I cannot give you a strong opinion about all this, but from what I've seen till now it is difficult for the newbie to have a clear picture of what methods should I use and when, between all the available ones.

I agree that p.b and p.h are the shortest and more pythonic variants, so I also prefer using them if it is possible!, but it would be ideal to have a clean guideline of the prefered usage of each method. Something like:
* If you change the tree structure (add/remove nodes, move them, clones, etc.) you should be using these methods: ...
* If you just edit a headline contents, you should use these other methods: ...
* You can safely **NOT** call p.setDirty() if you ....

and so on.

For example, in your sentence above:
> Similarly, p.setDirty() is *much* slower than p.v.setDirty(), but is often necessary.

it would be great for me to know why "is often necessary" and which are the main differences. And the [API docs](http://leo-editor.readthedocs.io/en/latest/index.html) are not of too much help... As a sample, the current documentation for the [p.b](http://leo-editor.readthedocs.io/en/latest/leo.core.html#leo.core.leoNodes.Position.b) method says just:
> position body string property

Of course, I can directly read the code, but it will be much harder for any one like me to get the general idea.

**Comment by edreamleo** [comment]
​​
On Mon, Apr 3, 2017 at 2:15 PM, XGiD <notifications@github.com> wrote:

> I agree that p.b and p.h are the shortest and more pythonic variants, so I
> also prefer using them if it is possible!, but it would be ideal to have a
> clean guideline of the prefered usage of each method. Something like:
>
​Thanks for your comments. ​

​Earlier today ​
​I chose to leave the API just as it is, and to improve the documentation
as you suggest.  I'll close this issue once the documentation is in place.
This reply will be pre-writing for the new docs.

For almost 35 years I have talked to my friend Phil Straus on Monday
mornings.  Maybe 30 times a year.  We talk about our goals for the coming
week, share brags of the week, and generally discuss what seems important
to us.

In our conversation today I mentioned that it's not really possible to
remove the imperfections mentioned in this issue.  This is a case of
perfectionism not being our friend.  Yes, the p.b and p.h setters, and
p.setDirty have performance gotchas, but ​the alternatives are worse.

Otoh, all fit and finish issues are imperfections that truly *must* be
fixed because they undermine the perceived quality of Leo.

Here's an answer to your implied questions:

The performance gotchas arise because these methods/properties update Leo's
appearance. Calling c.redraw() does *not *suffice.

For example, p.setDirty() not only sets p's headline dirty (redrawing the
*only* the icon), but also sets all ancestor @file nodes dirty, redrawing
*their* icons.  Yes, c.redraw() could do that calculation, but moving that
task from p.setDirty to c.redraw would impose a new requirement on *all *code
(including existing code), namely that all code that presently calls
p.setDirty() now must call c.redraw() later.  In general, such global
linkages between methods are far worse than performance gotchas.

Happily, these gotchas only become serious for repetitive commands, like
cff, replace-all and recursive import. In these situations, the code should
use p.v.b and p.v.h setters instead of p.b and p.h setters. Calling
p.setDirty() repetitively is less serious in practice. That's about all you
need to know.

Edward

**Comment by edreamleo** [comment]
> This reply will be pre-writing for the new docs.

Here is the first draft of the new docs:

**Warning**: The p.b and p.h setters and p.setDirty() are *very* expensive:

- p.b = s calls c.setBodyString(p, s) which will recolor body text and update the node's icon.
- p.h = s calls c.setHeadString(p, s) which calls p.setDirty().
- p.setDirty() changes the icons of all ancestor `@file` nodes.

In contrast, the corresponding p.v.b and p.v.b setters and p.v.setDirty() are extremely fast.

Usually, code *should* use p.setDirty() and the p.b and p.h setters, despite their cost, because they update Leo's outline pane properly. Calling c.redraw() is *not* enough.

These performance gotchas become important for repetitive commands, like cff, replace-all and recursive import. In such situations, code should use p.v.b and p.v.h setters instead of p.b and p.h setters.

**Comment by edreamleo** [comment]
76fb82a through c415cd4 document the gotchas in the appropriate docstrings, and in CheatSheet.leo.
